### PR TITLE
backend/azurerm: support for custom resource manager endpoints

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This allows the AzureRM Backend to be used with Azure Stack

Needs to be merged after #19448

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Tests pass across Azure Germany, Azure Public and Azure Stack:

screenshot 2018-11-26 at 12 14 54

File: backend/remote-state/azure/arm_client.go [hunk 1]
```diff
@@ -93,6 +95,16 @@ func buildArmClient(config BackendConfig) (*ArmClient, error) {
 	return &client, nil
 }
 
+func buildArmEnvironment(config BackendConfig) (*azure.Environment, error) {
+	if config.CustomResourceManagerEndpoint != "" {
+		log.Printf("Loading Environment from Endpoint %q", config.CustomResourceManagerEndpoint)
+		return authentication.LoadEnvironmentFromUrl(config.CustomResourceManagerEndpoint)
+	}
+
+	log.Printf("Loading Environment %q", config.Environment)
```
Comment (radeksimko) [inline backend/remote-state/azure/arm_client.go]: I think these log messages need the `[SEVERITY]` prefix. 👀

File: backend/remote-state/azure/helpers_test.go [hunk 2]
```diff
@@ -16,7 +16,8 @@ import (
 )
 
 const (
-	sasSignedVersion = "2017-07-29"
+	// required for Azure Stack
+	sasSignedVersion = "2015-04-05"
```
Comment (radeksimko) [inline backend/remote-state/azure/helpers_test.go]: Isn't this a breaking change? 🤔

File: backend/remote-state/azure/helpers_test.go [hunk 3]
```diff
@@ -99,7 +103,9 @@ func buildSasToken(accountName, accessKey string) (*string, error) {
 	signedVersion := sasSignedVersion
 
 	utcNow := time.Now().UTC()
-	startDate := utcNow.Format(time.RFC3339)
+
+	// account for servers being up to 5 minutes out
+	startDate := utcNow.Add(time.Minute * -5).Format(time.RFC3339)
```
Comment (radeksimko) [inline backend/remote-state/azure/helpers_test.go]: 🙈 🤦‍♂️

File: backend/remote-state/azure/helpers_test.go (hunk 2 above)
Comment (tombuildsstuff) [reply backend/remote-state/azure/helpers_test.go]: this feature hasn't shipped yet - but this is a test either way (we're assuming users submit a valid SAS Token)

File: backend/remote-state/azure/helpers_test.go (hunk 3 above)
Comment (tombuildsstuff) [reply backend/remote-state/azure/helpers_test.go]: 🙃

**Pull request merged**